### ISSUE TITLE
Container changes in`Deployment` => "element [0] missing"

### ISSUE THREAD

**Issue by Dominik-K** [weaveworks/kubediff#37]
Using this `deployment.yaml`:

    apiVersion: extensions/v1beta1
    kind: Deployment
    metadata:
      name: nginx
      namespace: test
    spec:
      template:
        metadata:
          labels:
            app: nginx
        spec:
          containers:
            - name: nginx
              image: nginx
              ports:
                - containerPort: 80
                  name: http
                  protocol: TCP

If I change a field in the `containers` part, e.g. the `containerPort`, `kubediff` returns:

    ## test/nginx (Deployment)

    .spec.template.spec.containers: 'element [0]' missing

**Comment by tomwilkie** [comment]
Yeah, thats an unfortunate side effect of https://github.com/weaveworks/kubediff/pull/31.

@awh can we improve this?

**Comment by awh** [comment]
We can, by teaching kubediff to sort certain lists e.g. `.spec.template.spec.containers` should be sorted by `.name` before comparison.

**Comment by ademariag** [comment]
is there an update on this? or a workaround?

**Comment by damienleger** [comment]
I've coded a workaround for this. It's far from perfect because tests are broken and all but that fits my needs. Looks interesting to share anyway.

My understanding is that after the change https://github.com/weaveworks/kubediff/pull/31 every change in any deep after a list results in 'element[X] missing' issue we are talking about. Hence if you change the image of a container (the most likely on a daily change), you have this issue, making the tool quite useless imho. The change make sense in a way that if you remove or add element in a list from running or wished configuration, differences between elements get mixed up and you end up with an horrible diff output e.g.: you remove wished var1 then running var1 is diffed with wished var2, running var2 is diffed with wished var3 and so on... Luckily changes not happening much except for environment variable list.

I then did the following change. Reverted the old behavior for list having same size, and if not return 'element missing' except for .spec.template.spec.containers[X].env whish have a special treat.

Let's say you running on this

`
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
  namespace: test
spec:
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx:1.14.1
          ports:
            - containerPort: 80
              name: http
              protocol: TCP
          env:
            - name: name1
              value: value1
            - name: name2
              value: value2
            - name: name3
              value: value3

`

and want this:

`
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
  namespace: test
spec:
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx:1.14.2
          ports:
            - containerPort: 80
              name: http
              protocol: TCP
          env:
            - name: name1
              value: value1.1
            - name: name4
              value: value4

`

With my hack, kubediff output is:

`
## test/nginx (Deployment.v1beta1.extensions)

.spec.template.spec.containers[0].image: 'nginx:1.14.2' != 'nginx:1.14.1'
## test/nginx (Deployment.v1beta1.extensions)

.spec.template.spec.containers[0].env: Unequal lengths: 2 != 3
## test/nginx (Deployment.v1beta1.extensions)

.spec.template.spec.containers[0].env: Diff: name4 in wished config only
## test/nginx (Deployment.v1beta1.extensions)

.spec.template.spec.containers[0].env: Diff: name2 in running config only
## test/nginx (Deployment.v1beta1.extensions)

.spec.template.spec.containers[0].env: Diff: name3 in running config only
## test/nginx (Deployment.v1beta1.extensions)

.spec.template.spec.containers[0].env[0].value: 'value1.1' != 'value1'

`

Patch is attached. Edit kubedifflib/_diff.py and build container with build instructions in doc except "make .uptodate" instead of "make" to bypass tests.

[patch.txt](https://github.com/weaveworks/kubediff/files/2859894/patch.txt)

**Comment by collinbachi** [comment]
Any update here? I have to imagine comparing changes to containers is pretty crucial to most people's use cases.

**Comment by chancez** [comment]
Trying kubediff for the first time and I get this on an array inside one of the values for my HelmRelease.

Eg:

```
kubediff consul grafana loki prometheus-k8s vault
## loki/loki (HelmRelease.v1beta1.flux.weave.works)

.spec.values.loki.config.schema_config.configs: 'element [0]' missing
```

With the following HelmRelease:

```

**Comment by Smana** [comment]
The patch from @damienleger works great. thanks

**Comment by damienleger** [comment]
hello @gouravsw, tbh I've did that tweak for my previous job so I'm not working on this anymore. That being said, when referring to "doc" I meant this --> https://github.com/weaveworks/kubediff#build

So by replacing `make` by `make .uptodate`, here it is:

```
mkdir -p $GOPATH/src/github.com/prometheus && cd "$_"
git clone git@github.com:prometheus/client_golang.git
mkdir -p $GOPATH/src/github.com/weaveworks && cd "$_"
git clone git@github.com:weaveworks/kubediff.git
cd kubediff
make
```
